subprojects/core-api/src/main/java/org/gradle/api/attributes/CompileView.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.attributes;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.Named;
+
+/**
+ * Differentiates between the different views of a library's compile classpath.
+ *
+ * <p>This attribute is only applicable when paired with the {@link Usage#JAVA_API} attribute.</p>
```

Comment:
Whatever solution we go with, we want to be sure that during compile-time we get this new variant, and during runtime we get the `runtimeElements` variant. 

The plan is for the tests to add a `project()` dependency with this attribute on the dependency. This gets merged with the attributes on the configuration doing the resolution.

So, when resolving the compile classpath, we get:
`Usage=java-api` (from the compile classpath attributes)
`CompileView=java-implementation` (from dependency attributes)

This selects the new variant

And when resolving the runtime classpath we get:
`Usage=java-runtime` (from runtime classpath attributes)
`CompileView=java-implementation` (from dependency attribute)

This selects `runtimeElements`

Now, if we instead did not add a new attribute but added a new `Usage`:

When resolving the compile classpath:
`Usage=java-api` from configuration is overridden
`Usage=java-implementation` (from dependency attributes)

This selects the new variant

When resolving the runtime classpath:
`Usage=java-runtime` from configuration is overridden
`Usage=java-implementation` (from dependency attributes)

This selects the new variant

I am open to other ideas, but this was the major driving factor in adding a new attribute

Reply:
Ok, that makes sense to me now.  I was assuming that this would be set on the consuming configuration rather than the dependency.  But I guess that wouldn't make sense because then we might get implementation dependencies of other projects when we only want it for the current project.  I'm catching up.